services/autoscaling/src/simcore_service_autoscaling/modules/instrumentation.py

```diff
@@ -0,0 +1,125 @@
+from dataclasses import dataclass, field
+from typing import Final, cast
+
+from fastapi import FastAPI
+from prometheus_client import CollectorRegistry, Counter, Gauge
+from servicelib.fastapi.prometheus_instrumentation import (
+    setup_prometheus_instrumentation as setup_rest_instrumentation,
+)
+
+from ..core.errors import ConfigurationError
+from ..core.settings import get_application_settings
+from ..models import AssociatedInstance, Cluster, NonAssociatedInstance
+
+EC2_INSTANCE_LABELS: Final[tuple[str]] = ("instance_type",)
+
+
+def _update_gauge(
+    gauge: Gauge,
+    nodes: list[AssociatedInstance] | list[NonAssociatedInstance],
+) -> None:
+    gauge.clear()
+
+    for node in nodes:
+        gauge.labels(instance_type=node.ec2_instance.type).inc()
+
+
+@dataclass(slots=True, kw_only=True)
+class AutoscalingInstrumentation:
+    registry: CollectorRegistry
+    _active_nodes: Gauge = field(init=False)
+    _pending_nodes: Gauge = field(init=False)
+    _drained_nodes: Gauge = field(init=False)
+    _buffer_drained_nodes: Gauge = field(init=False)
+    _pending_ec2s: Gauge = field(init=False)
+    _disconnected_nodes: Gauge = field(init=False)
+    _started_instances: Counter = field(init=False)
+    _terminated_instances: Counter = field(init=False)
+
+    def __post_init__(self) -> None:
+        self._active_nodes = Gauge(
+            "active_nodes",
+            "Number of EC2-backed docker nodes which are active and ready to run tasks",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._pending_nodes = Gauge(
+            "pending_nodes",
+            "Number of EC2-backed docker nodes which are active and NOT ready to run tasks",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._drained_nodes = Gauge(
+            "drained_nodes",
+            "Number of EC2-backed docker nodes which are drained",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._buffer_drained_nodes = Gauge(
+            "buffer_drained_nodes",
+            "Number of EC2-backed docker nodes which are drained and in buffer/reserve",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._pending_ec2s = Gauge(
+            "pending_ec2s",
+            "Number of EC2 instance not yet part of the cluster",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._disconnected_nodes = Gauge(
+            "disconnected_nodes",
+            "Number of docker node not backed by a running EC2 instance",
+        )
+        self._started_instances = Counter(
```

Comment:
Why did you use a Counter here? As far as I undersand , this number can only increase. So this would represent how many instances we have started in the entire history?

SEE https://prometheus.io/docs/concepts/metric_types/#counter

Reply:
yes. by definition it is a counter. started instances as the name imply can only increase, same as with terminated instances